Fix rmiregistry AccessControlException

JasonFengJ9 opened the pull request:
Fix `rmiregistry AccessControlException`

bootstrap classloader should load a class via `VMAccess.findClassOrNull()` instead of `ClassLoader.defineClass()`. `VMAccess.findClassOrNull()` will load class with all permissions while `ClassLoader.defineClass()` loads class with a protection domain within the security context.
Added a test for this scenario.

Issue: #301

Reviewer @pshipton 
FYI: @DanHeidinga 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

pshipton commented:
Please provide more detail about how and why this fixes the permission issue,

ClassLoader.forName(Module module, String name) also has a call to loadClassHelper() which doesn't delegate, does it also need changes?

pshipton commented on jcl/src/java.base/share/classes/java/lang/ClassLoader.java:
```diff
@@ -1206,7 +1206,7 @@ public final Module getUnnamedModule()
 	Class<?> localClass = null;
 	try {
-		localClass = loadClassHelper(className, false, false, module);
+		localClass = loadClassHelper(className, false, true, module);
```
The method comment needs updating to match.

JasonFengJ9 replied:
Updated.

JasonFengJ9 commented:
Without this PR, `ClassLoader.loadClass()` goes through following stackgtrace for class loading:
````
	at java.lang.ClassLoader.defineClass(java.base@9-internal/ClassLoader.java:450)
	at java.lang.ClassLoader.defineClass(java.base@9-internal/ClassLoader.java:2002)
	at java.security.SecureClassLoader.defineClass(java.base@9-internal/SecureClassLoader.java:206)
	at jdk.internal.loader.BuiltinClassLoader.defineClass(java.base@9-internal/BuiltinClassLoader.java:1021)
	at jdk.internal.loader.BuiltinClassLoader.findClassInModuleOrNull(java.base@9-internal/BuiltinClassLoader.java:862)
	at jdk.internal.loader.BuiltinClassLoader.findClass(java.base@9-internal/BuiltinClassLoader.java:733)
	at java.lang.ClassLoader.loadClassHelper(java.base@9-internal/ClassLoader.java:1148)
	at java.lang.ClassLoader.loadClass(java.base@9-internal/ClassLoader.java:1086)
	at java.lang.Class.forName(java.base@9-internal/Class.java:438)
	at sun.launcher.LauncherHelper.loadModuleMainClass(java.base@9-internal/LauncherHelper.java:611)
	at sun.launcher.LauncherHelper.checkAndLoadMain(java.base@9-internal/LauncherHelper.java:550)
````
`ClassLoader.defineClass` will associate a protection domain (not all permission) to the class to be defined, and cause AccessControlException when security checking is performed.

With this PR, the class in question is loaded as following stacktrace:
````
        native VMAccess.findClassOrNull()
	at java.lang.ClassLoader.loadClass(java.base@9-internal/ClassLoader.java:1018)
	at java.lang.ClassLoader.loadClassHelper(java.base@9-internal/ClassLoader.java:1106)
	at java.lang.ClassLoader.loadClass(java.base@9-internal/ClassLoader.java:1065)
	at java.lang.Class.forName(java.base@9-internal/Class.java:440)
	at sun.launcher.LauncherHelper.loadModuleMainClass(java.base@9-internal/LauncherHelper.java:611)
	at sun.launcher.LauncherHelper.checkAndLoadMain(java.base@9-internal/LauncherHelper.java:550)
````
This ensure the classes from the bootstrap modules are loaded via `native VMAccess.findClassOrNull` which has null security context hence with all permissions.

Agreed,  `ClassLoader.forName(Module module, String name)` shall invoke `loadClassHelper` with `delegateToParent` set to true. In general class loading need delegate to parent classloader unless it is specified to load local classes for current classloader.

pshipton commented:
We need new test cases added for this please, to check classes both loaded by the bootstrap classloader, and by the platform or application classloader.

JasonFengJ9 commented:
Sorry for late update due to other workload. 
Removed the change to `classLoader.loadClassHelper(name, false, false, module)` within `Class.forName(Module module, String name)`. This method performs a check whether loaded class belongs to same module as incoming module loaded by current classloader, if not, `null` will be returned. Hence there is no need to delegate to parent classloader which loads different modules. 

Added a testcase to validate that the change to `ClassLoader.loadClass(module, className)` ensures the parent classloader (bootstrap classloader in this case) is delegated to load the class first. Without this pull request, the testcase will fail, and reports `AllPermission()` wasn't granted to a bootstrap class just like `sun.rmi.registry.RegistryImpl` eventually it could lead to `AccessControlException` reported in the issue #301.

Pls review.

pshipton commented on test/Java9andUp/src_latest/org/openj9/test/java/lang/Test_Class.java:
```diff
@@ -0,0 +1,63 @@
+		}
+
+		Module  jrModule = jrAccessExceptionClz.getModule();
+		Class<?> srrRegistryImplClz = Class.forName(jrModule, "sun.rmi.registry.RegistryImpl");
```
It would be better to load a class from an exported package to avoid export problems in the future when the default access controls are changed.

JasonFengJ9 replied:
Will replace it with `java.rmi.AlreadyBoundException` instead.

pshipton commented:
The fix doesn't seem correct. As I understand it, the ClassLoader is already the bootstrapClassLoader, yet setting delegateToParent needs to be set to true in order to load correctly from the same ClassLoader. There is no actual delegation occurring, just different APIs are used to load correctly from the bootstrapClassLoader. It seems the same fix can be made without setting delegateToParent by checking for the bootstrapClassLoader and calling loadClass() instead of findClass().

JasonFengJ9 commented:
Removed the delegation flag change, added a check for bootstrap classloader within `j.l.ClassLoader.loadClass(module, className)` to invoke `VMAccess.findClassOrNull(className, bootstrapClassLoader)` for bootstrap classloader instead.
Pls have another look.

pshipton commented on jcl/src/java.base/share/classes/java/lang/ClassLoader.java:
```diff
@@ -1205,10 +1205,15 @@ public final Module getUnnamedModule()
-	} catch (ClassNotFoundException e) {
-		// returns null if the class can't be found
+	
+	if ((bootstrapClassLoader == null) || (this == bootstrapClassLoader)) {
```
Shouldn't the fix go into loadClassHelper() so other callers of loadClassHelper() get the same benefits?

JasonFengJ9 replied:
Do you mean moving this code snippet from `loadClass (className)` and `loadClass(module, className)` into `loadClassHelper()`?
`BOOT_LOADER` extends `BuiltinClassLoader` which overrides `loadClass(className, resolveClass)`, hence `BOOT_LOADER.loadClass (className)` will invoked this overridden method and missed the checking for bootstrap classloader within `loadClassHelper()` hence can't call `VMAccess.findClassOrNull(className, bootstrapClassLoader)` for bootstrap classes. I think this was the main reason we put this checking code snippet within `loadClass (className)`. This PR adds the check for `loadClass(module, className)` as well.
This is for bootstrap classloader class loading, other classloader use scenario doesn't need this check.

pshipton commented on test/Java9andUp/src_latest/org/openj9/test/java/lang/Test_Class.java:
```diff
@@ -0,0 +1,64 @@
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
+
+@Test(groups = { "level.extended" })
```
This seems an important test, and its also quick to run. I suggest adding it to sanity to avoid any breakage with future changes.

pshipton commented:
@llxia any concerns with the test changes?

llxia commented:
The test looks good to me.
If we want to add this test to sanity, then both annotation in Test_Class.java and tag in playlist need to be updated.

JasonFengJ9 commented:
Changed the test group to `sanity`. Also added the test in `src_current` besides `src_latest` such that both `OpenJ9 Java 9` and `Java 9 b148` can run the tests. (Actually it is required otherwise `b148` tests won't be able to find the classes).
@llxia pls have another look.
